dm-agent/src/main.rs

```diff
@@ -0,0 +1,190 @@
+    )?;
+
+    Ok(())
+}
+
+#[instrument(name = "message_batch_handler")]
+fn handle_message_batch(mqtt: Arc<Client>, message_batch: MeasurementGrouper) {
+    tokio::task::spawn(async move {
+        if let Err(err) = message_batch_processor(mqtt, message_batch).await {
+            error!("{}", err);
+        }
+    });
+}
+
+async fn message_batch_processor(
+    mqtt: Arc<Client>,
+    message_batch: MeasurementGrouper,
+) -> anyhow::Result<()> {
+    let mut tedge_json_serializer = ThinEdgeJsonSerializer::new();
+    message_batch.accept(&mut tedge_json_serializer)?;
+
+    let topic = Topic::new(TARGET_TOPIC).unwrap();
+    let tedge_message = Message::new(&topic, tedge_json_serializer.bytes()?);
+
+    mqtt.as_ref().publish(tedge_message).await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_message_batch_processor() {
+        let mut message_batch = MeasurementGrouper::new();
+
+        let topic = Topic::new("collectd/temperature/value").unwrap();
+        let collectd_message = Message::new(&topic, "123456789:32.5");
+        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();
+
+        let topic = Topic::new("collectd/coordinate/x").unwrap();
+        let collectd_message = Message::new(&topic, "123456789:50");
+        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();
+
+        let topic = Topic::new("collectd/coordinate/y").unwrap();
+        let collectd_message = Message::new(&topic, "123456789:70");
+        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();
+
+        let topic = Topic::new("collectd/pressure/value").unwrap();
+        let collectd_message = Message::new(&topic, "123456789:98.2");
+        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();
+
+        let topic = Topic::new("collectd/coordinate/z").unwrap();
+        let collectd_message = Message::new(&topic, "123456789:90");
+        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();
+
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("temperature"), "value")
+                .unwrap(),
+            32.5
+        );
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("pressure"), "value")
+                .unwrap(),
+            98.2
+        );
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("coordinate"), "x")
+                .unwrap(),
+            50.0
+        );
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("coordinate"), "y")
+                .unwrap(),
+            70.0
+        );
+        assert_eq!(
+            message_batch
+                .get_measurement_value(Some("coordinate"), "z")
+                .unwrap(),
+            90.0
+        );
+    }
+}
```

Comment:
Could we split that test into separate tests and name them properly, eg:

```rust
fn message_batch_processor_temparature_value_correct() {
        let topic = Topic::new("collectd/temperature/value").unwrap();
        let collectd_message = Message::new(&topic, "123456789:32.5");
        tedge_json_from_collectd_message(&mut message_batch, collectd_message).unwrap();

        assert_eq!(
            message_batch
                .get_measurement_value(Some("temperature"), "value")
                .unwrap(),
            32.5
        );
}

Reply:
Why do you think this needs splitting? The intent of the test is to see if the API under test can build a Thin Edge JSON message batch from multiple collectd messages, a mix of single valued and multi-valued measurements. Splitting will defeat that purpose. I could add more tests to see the handling of a batch with just one message. But not sure if that adds any value in the presence of this test.